compiler/test/dotty/tools/dotc/ParallelTesting.scala

```diff
@@ -0,0 +1,1076 @@
+   *
+   *  To verify output you may use `.check` files. These files should share the
+   *  name of the file or directory that they are testing. For instance:
+   *
+   *  ```none
+   *  .
+   *  └── tests
+   *      ├── i1513.scala
+   *      └── i1513.check
+   *  ```
+   *
+   *  If you are testing a directory under separate compilation, you would
+   *  have:
+   *
+   *  ```none
+   *  .
+   *  └── tests
+   *      ├── myTestDir
+   *      │   ├── T_1.scala
+   *      │   ├── T_2.scala
+   *      │   └── T_3.scala
+   *      └── myTestDir.check
+   *  ```
+   *
+   *  In the above example, `i1513.scala` and one of the files `T_X.scala`
+   *  would contain a `Test` object with a main method.
+   *
+   *  Composing tests
+   *  ===============
+   *  Since this is a parallel test suite, it is essential to be able to
+   *  compose tests to take advantage of the concurrency. This is done using
+   *  the `+` function. This function will make sure that tests being combined
+   *  are compatible.
+   */
+  final class CompilationTest private (
+    private[ParallelTesting] val targets: List[Target],
+    private[ParallelTesting] val times: Int,
+    private[ParallelTesting] val shouldDelete: Boolean,
+    private[ParallelTesting] val threadLimit: Option[Int],
+    private[ParallelTesting] val shouldFail: Boolean
+  ) {
+    import org.junit.Assert.fail
+
+    private[ParallelTesting] def this(target: Target) =
+      this(List(target), 1, true, None, false)
+
+    private[ParallelTesting] def this(targets: List[Target]) =
+      this(targets, 1, true, None, false)
+
+    /** Compose test targets from `this` with `other`
+     *
+     *  It does this, only if the two tests are compatible. Otherwise it throws
+     *  an `IllegalArgumentException`.
+     *
+     *  Grouping tests together like this allows us to take advantage of the
+     *  concurrency offered by this test suite as each call to an executing
+     *  method (`pos()` / `neg()`/ `run()`) will spin up a thread pool with the
+     *  maximum allowed level of concurrency. Doing this for only a few targets
+     *  does not yield any real benefit over sequential compilation.
+     *
+     *  As such, each `CompilationTest` should contain as many targets as
+     *  possible.
+     */
+    def +(other: CompilationTest) = {
+      require(other.times == times, "can't combine tests that are meant to be benchmark compiled")
+      require(other.shouldDelete == shouldDelete, "can't combine tests that differ on deleting output")
+      require(other.shouldFail == shouldFail, "can't combine tests that have different expectations on outcome")
+      new CompilationTest(targets ++ other.targets, times, shouldDelete, threadLimit, shouldFail)
+    }
+
+    /** Creates a "pos" test run, which makes sure that all tests pass
+     *  compilation without generating errors and that they do not crash the
+     *  compiler
+     */
+    def pos(): this.type = {
```

Comment:
pos is a very bad name for such an important method.
I'd call it "checkCompiles"